vllm/entrypoints/openai/serving_responses.py

```diff
@@ -748,25 +751,36 @@ def _construct_input_messages_with_harmony(
     async def _run_background_request(
         self,
         request: ResponsesRequest,
+        context: ConversationContext,
         *args,
         **kwargs,
     ):
-        try:
-            response = await self.responses_full_generator(
-                request, *args, **kwargs)
-        except Exception as e:
-            logger.exception("Background request failed for %s",
-                             request.request_id)
-            response = self.create_error_response(str(e))
-
-        if isinstance(response, ErrorResponse):
-            # If the request has failed, update the status to "failed".
-            response_id = request.request_id
-            async with self.response_store_lock:
-                stored_response = self.response_store.get(response_id)
-                assert stored_response is not None
-                if stored_response.status not in ("completed", "cancelled"):
-                    stored_response.status = "failed"
+        async with AsyncExitStack() as exit_stack:
+            try:
+                if self.tool_server is not None:
+                    tool_list = context.list_lazy_init_tools()
+                    for tool_name in tool_list:
+                        tool_session_ctx = exit_stack.enter_async_context(
+                            self.tool_server.new_session(tool_name))
+                        context.tool_sessions[
+                            tool_name] = await tool_session_ctx
```

Comment:
prefer to wrap to logic into something like `context.init_tool_sessions(tool_server, exist_stack)`